sigle/src/modules/editor/extensions/Twitter/TwitterComponent.tsx

```diff
@@ -0,0 +1,334 @@
+  const [tweetCreated, setTweetCreated] = useState(false);
+
+  const tweetId = props.node.attrs['data-twitter-id'];
+  const tweetUrl = props.node.attrs.url;
+  const isPasted = props.node.attrs.pasted;
+
+  useEffect(() => {
+    if (!tweetId && inputRef) {
+      // override tiptap focus
+      setTimeout(() => {
+        inputRef.current?.focus();
+      }, 5);
+    }
+  }, []);
+
+  useEffect(() => {
+    if (tweetId) {
+      loadTwitterWidget().then(() => {
+        createTweetOnLoad();
+      });
+    }
+  }, [tweetId]);
+
+  const formik = useFormik<TweetValues>({
+    initialValues: {
+      tweetUrl: tweetUrl || '',
+    },
+    validate: (values) => {
+      const errors: FormikErrors<TweetValues> = {};
+
+      if (!values.tweetUrl.match(TWITTER_REGEX)) {
+        errors.tweetUrl = 'Invalid URL';
+      }
+
+      return errors;
+    },
+    validateOnBlur: false,
+    validateOnChange: false,
+    onSubmit: async (values, { validateForm }) => {
+      validateForm().then(() => {
+        formik.setSubmitting(true);
+
+        props.editor.commands.updateAttributes('twitter', {
+          ...props.node.attrs,
+          ['data-twitter-id']: getTweetIdFromUrl(values.tweetUrl),
+        });
+      });
+
+      submitTweetId();
+    },
+  });
+
+  const createTweetOnLoad = async () => {
+    setIsTweetLoading(true);
+    createTweet(tweetId, containerRef).then((value: any) => {
+      if (!value) {
+        formik.setErrors({ tweetUrl: 'Create tweet error' });
+        setIsTweetLoading(false);
+        return;
+      }
+      if (isPasted) {
+        props.editor.commands.updateAttributes('twitter', {
+          ...props.node.attrs,
+          pasted: false,
+        });
+        props.editor.commands.createParagraphNear();
+      }
+      setTweetCreated(true);
+      setIsTweetLoading(false);
+    });
+  };
+
+  const submitTweetId = () => {
+    // clear errors if retrying submit
+    if (formik.errors) {
+      formik.setErrors({ tweetUrl: undefined });
+    }
+
+    // if retrying to submit after an error use existing tweet id rather than running function on each retry
+    const id = tweetId || getTweetIdFromUrl(formik.values.tweetUrl);
+
+    // @ts-expect-error Twitter is attached to the window.
+    if (!id || !window.twttr) {
+      return;
+    }
+
+    createTweet(id, containerRef).then((value) => {
+      if (!value) {
+        formik.setErrors({ tweetUrl: 'Create tweet error' });
+        formik.setSubmitting(false);
+        return;
+      }
+
+      props.editor.commands.createParagraphNear();
+      setTweetCreated(true);
+      formik.setSubmitting(false);
+    });
+  };
+
+  const onKeyDown = (event: React.KeyboardEvent) => {
+    // Remove input if empty and user presses backspace or delete key
+    if (
+      (!formik.values.tweetUrl && event.key === 'Backspace') ||
+      event.key === 'Delete'
+    ) {
+      props.deleteNode();
+    }
+  };
+
+  const pasteUrlAsLink = () => {
+    const editor = props.editor;
```

Comment:
here we should check that it's a valid link otherwise insert it as text

Reply:
we have the check for the valid link at line 133, or is there a reason we should still validate it here?

Reply:
if you enter in the input any sentence which is not a link eg: "Tokenomics Research @ Stacks Foundation | Author 'Tokenomics for Builders'" and then click "Paste URL as link", it's creating an invalid link :)

Reply:
ahh yes makes sense, will update it :)